### PR TITLE
Hint registration

### LINKED ISSUE #6627: SDL hints as Quake-style cvars
> (This is brought on by #6581 finding out SDL_GetVersion now takes 16 milliseconds.)
>
> Idea: the Hint subsystem should work like quake cvars.
>
> Instead of a callback and a generic SDL_GetHint() function, you do something like this:
>
> ```c
> static SDL_bool support_foobars = SDL_TRUE;
> SDL_RegisterHintBool("SDL_SUPPORT_FOOBARS", &support_foobars);
> ```
>
> This adds this to a list or hash for later use, and sets a default if an environment variable is already set. When someone calls SDL_SetHint("SDL_SUPPORT_FOOBARS", "1"), it updates the variable directly. If a hint is set before it's registered, we'll keep track of it and RegisterHint will work out the bookkeeping like it does for environment variables. One can also mark a hint as locked (once we use it, we don't want the value to change)
>
> Now when SDL needs to check a hint, it just looks at a static variable instead of calling a function, and you don't need a callback just to update one. The end result is that hint lookups are always _extremely_ fast and take less effort to maintain individually, as the complexity moves to a single common piece of code (in the set and register functions).
>
> We can still keep an optional callback mechanism if there's work that needs to be done when a variable is set, but I'm more inclined to say there should be a simple "this has changed since the last time you used it" flag that code can check before using the value.
>
> SDL_GetHint() can remain, in case a hint needs to be looked up by name, but internally, SDL can just look at variables internally.

_Originally posted by @icculus in https://github.com/libsdl-org/SDL/issues/3519#issuecomment-1324206292_

### PR DESCRIPTION
**Pull request by icculus**
This PR adds hint "registration," which basically makes hints that have simple requirements easier to manage.

Basically you register a hint to an address, and it'll update that address whenever the hint was set. This piggybacks off the existing hint callback mechanism, so largely this cleans up existing code.

I originally intended to rip out the existing hint subsystem and turn it into something more like how Quake handles its "cvars," but it turns out the existing code is pretty powerful, and the specific feature I desired is easy to layer onto the existing callbacks with very little new code and without adding any new allocations...and we can pick and choose what makes sense--between SDL_GetHint, callbacks, and the new registration interface--on a per-hint basis.

Note that this is (currently) an internal-only interface.

I've cleaned out all the callbacks that are just doing simple variable upkeep, but I haven't started evaluating any of the internal calls to SDL_GetHint. Most of those should be cleaned out, too, as it will turn a pile of string comparisons while traversing a linked list, plus a getenv() call, plus parsing a string, into a simple static variable access.

This is still a draft PR until I do that.

Note that this could probably be applied to SDL 2.28.0 as well, assuming other factors haven't made this difficult to merge.

Fixes #6627.

### CODE REVIEW

File: src/SDL_hints.c [hunk 1]
```diff
@@ -43,6 +47,7 @@ typedef struct SDL_Hint
 
 static SDL_Hint *SDL_hints;
 
+/* !!! FIXME: should we set the SDL error string in here for failures? */
```
Comment (slouken) [inline src/SDL_hints.c:50]: Yes. :)

**Comment by slouken** [comment]
This is an awesome change! :)
I don't think we need to backport this to SDL2, I'm guessing we won't add a lot of new code that uses hints.

Food for thought... should the callback in SDL3 be called _after_ the hint value has changed? We pass the old value to the callback, and there's a bunch of places I've wanted to look up the hint that has changed as part of updating a multi-hint state.

**Comment by slouken** [comment]
We should definitely expose these in the public API for SDL3, they seem super useful for the simple case.

File: src/SDL_hints.c (hunk 1 above)
Comment (icculus) [reply src/SDL_hints.c]: Fixed!

File: src/SDL_hints.c [hunk 2]
```diff
@@ -296,3 +301,77 @@ void SDL_ClearHints(void)
...
+static void SDLCALL UpdateRegisteredFloatHintCallback(void *userdata, const char *name, const char *oldValue, const char *newValue)
+{
+    *((float *) userdata) = newValue ? SDL_atof(newValue) : 0.0f;
+}
+
+SDL_bool SDL_RegisterFloatHint(const char *name, float *addr, float default_value)
+{
+    char defaultstr[64];
+    SDL_snprintf(defaultstr, sizeof (defaultstr), "%f", default_value);
+    (void) SDL_SetHintWithPriority(name, defaultstr, SDL_HINT_DEFAULT);  /* carry on even if this fails. */
```
[the remark this answers is not in the document]
Comment (icculus) [reply src/SDL_hints.c:353]: Hmm, this is a good question, resetting the hint will set them to NULL, which will cause them to be zeroed out. We would need to store the default in the hint structure to handle this case.

[the remark this answers is not in the document]
Comment (slouken) [reply src/SDL_hints.c:353]: I wonder if we should store the default value in the hint directly, and then use that in addition to the environment when looking up the default or resetting the hint?

[the remark this answers is not in the document]
Comment (slouken) [reply src/SDL_hints.c:353]: Great minds think alike. :)

**Comment by icculus** [comment]
> Food for thought... should the callback in SDL3 be called _after_ the hint value has changed? We pass the old value to the callback, and there's a bunch of places I've wanted to look up the hint that has changed as part of updating a multi-hint state.

This is fine with me, I think.

File: src/SDL_hints.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (icculus) [reply src/SDL_hints.c]: This is resolved by https://github.com/libsdl-org/SDL/pull/7040/commits/835f8b67b2f77148d58609af60aaa4fcfc42a843 but this code feels a little hairy, if you want to look it over.

**Comment by icculus** [comment]
So I closed the associated issue as WONTFIX because this approach can't be made to be thread safe, but I'll leave this PR open and in the milestone until there's a moment to see if anything useful is salvageable from here.

Which it might not be, in which case, it's 100% okay to just close this without merging.

**Comment by icculus** [comment]
So we're going to leave this alone, because SDL_Properties has superceded this, and hints at this point are largely meant to either be set once at startup or SDL internally takes steps to protect individual hint strings that might change. So the thread safety issues are Good Enough for now and we'll let this die.

**Pull request closed without merging**